src/daft-schema/src/dtype.rs

```diff
@@ -659,6 +755,119 @@ impl DataType {
+    /// For example, `FixedSizeBinary` and `FixedSizeList` have a size property.
+    pub fn fixed_size(&self) -> Option<usize> {
+        match self {
+            Self::FixedSizeBinary(size) => Some(*size),
+            Self::FixedSizeList(_, size) => Some(*size),
+            Self::Embedding(_, size) => Some(*size),
+            _ => None,
+        }
+    }
+    /// if the datatype variant has a shape, return it.
+    /// For example, `FixedShapeImage` and `FixedShapeTensor` have a fixed shape.
+    pub fn fixed_shape(&self) -> Option<Vec<u64>> {
+        match self {
+            Self::FixedShapeImage(_, height, width) => Some(vec![*height as u64, *width as u64]),
+            Self::FixedShapeTensor(_, shape) => Some(shape.clone()),
+            Self::FixedShapeSparseTensor(_, shape, _) => Some(shape.clone()),
+            _ => None,
+        }
+    }
+
+    /// if the datatype contains a timeunit, return it.
+    pub fn time_unit(&self) -> Option<TimeUnit> {
+        match self {
+            Self::Timestamp(unit, _) => Some(*unit),
+            Self::Time(unit) => Some(*unit),
+            Self::Duration(unit) => Some(*unit),
+            _ => None,
+        }
+    }
+
+    /// if the datatype contains a timezone, return it.
+    pub fn time_zone(&self) -> Option<Option<&str>> {
+        match self {
+            Self::Timestamp(_, timezone) => Some(timezone.as_deref()),
+            _ => None,
+        }
+    }
+
+    /// if the datatype contains an image mode, return it.
+    /// For example, `Image` and `FixedShapeImage` have an image mode.
+    pub fn image_mode(&self) -> Option<ImageMode> {
+        match self {
+            Self::Image(mode) => *mode,
+            Self::FixedShapeImage(mode, ..) => Some(*mode),
+            _ => None,
+        }
+    }
+
+    /// if the datatype contains an inner datatype, return it.
+    /// For example, `List` and `FixedSizeList` have an inner datatype.
+    pub fn dtype(&self) -> Option<&Self> {
+        match self {
+            Self::List(dtype) | Self::FixedSizeList(dtype, _) => Some(dtype),
+            Self::Embedding(dtype, _) => Some(dtype),
+            Self::Extension(_, dtype, _) => Some(dtype),
+            Self::Tensor(dtype) => Some(dtype),
+            Self::SparseTensor(dtype, _) => Some(dtype),
+            Self::FixedShapeTensor(dtype, _) => Some(dtype),
+            Self::FixedShapeSparseTensor(dtype, _, _) => Some(dtype),
+            _ => None,
+        }
+    }
+
+    /// if the datatype is a struct, return its fields.
+    pub fn fields(&self) -> Option<&[Field]> {
+        match self {
+            Self::Struct(fields) => Some(fields),
+            _ => None,
+        }
+    }
+
+    /// if the datatype is a decimal, return its precision.
+    pub fn precision(&self) -> Option<usize> {
+        match self {
+            Self::Decimal128(precision, _) => Some(*precision),
+            _ => None,
+        }
+    }
```

Comment:
What are your thoughts on using `Result<Option<usize>>` since we may want to distinguish between the absence of a value vs a type-mismatch? For example, ImageMode is an optional parameter of Image, and we are not able to distinguish if the type is being used correctly vs the parameter is absent. I'm curious if you know of any other rust examples using sums with properties and how they handle this.

Reply:
we could switch to `Result<T>` instead of `Option<T>`. I think that might make things more intuitive for things like `timezone` which is an option already. `Option<Option<String>>` is kinda weird IMO

Reply:
What about `Result<Option<T>>`?

Reply:
* `Ok(Some(T))` - valid attribute, some value T
* `Ok(None)` - valid attribute, no value
* `Err` - invalid attribute

This let's us distinguish between an optional property and an invalid property like in the case of image_mode which is erroring now if not set, when it's valid for it to not be set. Timezone is another good example.

Reply:
Why would you want `Result<Option<usize>>` on something like precision? why not mirror the inner datatype representation and only use option when the field is optional.

`precision` -> `Result<usize>`
`image_mode` -> `Result<Option<ImageMode>>`
`time_zone` -> `Result<Option<String>>` 

What value add is there by wrapping something like `precision` in a `Option`? Keep in mind, these are not public apis, so I don't think we need to try to overcomplicate things.

Reply:
For precision, we don't need to wrap. The point was to use Result rather than Option to indicate whether or not the attribute is valid, which now let's us use options when the value is .. optional! 🙌 

As a fun aside, daft precision and scale are required for now, but not all systems require them funny enough! Arbitrary scale exact numbers are helpful when working with things like JSON.